src/adc.rs

```diff
@@ -0,0 +1,562 @@
+                        15 => self.rb.smpr2.modify(|_, w| w.smp15().bits(smp.bitcode())),
+                        16 => self.rb.smpr2.modify(|_, w| w.smp16().bits(smp.bitcode())),
+                        17 => self.rb.smpr2.modify(|_, w| w.smp17().bits(smp.bitcode())),
+                        18 => self.rb.smpr2.modify(|_, w| w.smp18().bits(smp.bitcode())),
+                        _ => unreachable!(),
+                    };
+                }
+
+            }
+
+            impl<WORD, PIN> OneShot<$ADC, WORD, PIN> for Adc<$ADC>
+            where
+                WORD: From<u16>,
+                PIN: Channel<$ADC, ID = u8>,
+                {
+                    type Error = ();
+
+                    fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
+                        let res = self.convert_one(PIN::channel());
+                        return Ok(res.into());
+                    }
+                }
+        )+
+    }
+}
+
+/// Macro to implement ADC functionallity for ADC1 and ADC2
+///
+/// TODO: Extend/differentiate beyond f303.
+macro_rules! adc12_hal {
+    ($(
+            $ADC:ident: ($adcx:ident),
+    )+) => {
+        $(
+            impl Adc<$ADC> {
+                fn enable_clock(&self, ahb: &mut AHB, adc_common: &mut ADC1_2) -> bool {
+                    if ahb.enr().read().adc34en().is_enabled() &&
+                        (adc_common.ccr.read().ckmode().variant() != self.ckmode.into()) {
+                        return false;
+                    }
+                    ahb.enr().modify(|_, w| w.adc12en().enabled());
+                    adc_common.ccr.modify(|_, w| w
+                        .ckmode().variant(self.ckmode.into())
+                    );
+                    true
+                }
+            }
+            adc_hal! {
+                $ADC: ($adcx, ADC1_2),
+            }
+        )+
+    }
+}
+
+/// Macro to implement ADC functionallity for ADC3 and ADC4
+///
+/// TODO: Extend/differentiate beyond f303.
+#[cfg(any(
+    feature = "stm32f303xb",
+    feature = "stm32f303xc",
+    feature = "stm32f303xd",
+    feature = "stm32f303xe",
+))]
+macro_rules! adc34_hal {
+    ($(
+            $ADC:ident: ($adcx:ident),
+    )+) => {
+        $(
+            impl Adc<$ADC> {
+                /// Returns true iff
+                ///     the clock was enabled
+                ///  or the clock was already enabled with the same settings
+                fn enable_clock(&self, ahb: &mut AHB, adc_common: &mut ADC3_4) -> bool {
+                    if ahb.enr().read().adc34en().is_enabled() {
+                        return (adc_common.ccr.read().ckmode().variant() == self.ckmode.into())
+                    }
+                    ahb.enr().modify(|_, w| w.adc34en().enabled());
+                    adc_common.ccr.modify(|_, w| w
+                        .ckmode().variant(self.ckmode.into())
+                    );
+                    true
+                }
```

Comment:
For consistency it would be nice if this function and the equivalent in `adc12_hal!` would have the same formatting.